// @sdeverywhere/runtime-async/dist/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  exposeModelWorker: () => exposeModelWorker,
  spawnAsyncModelRunner: () => spawnAsyncModelRunner
});
module.exports = __toCommonJS(src_exports);

// src/runner.ts
var import_threads = require("threads");
var import_runtime = require("@sdeverywhere/runtime");
async function spawnAsyncModelRunner(workerSpec) {
  if (workerSpec["path"]) {
    return spawnAsyncModelRunnerWithWorker(new import_threads.Worker(workerSpec["path"]));
  } else {
    return spawnAsyncModelRunnerWithWorker(import_threads.BlobWorker.fromText(workerSpec["source"]));
  }
}
async function spawnAsyncModelRunnerWithWorker(worker) {
  const modelWorker2 = await (0, import_threads.spawn)(worker);
  const initResult = await modelWorker2.initModel();
  const modelListing = initResult.modelListing ? new import_runtime.ModelListing(initResult.modelListing) : void 0;
  const params2 = new import_runtime.BufferedRunModelParams(modelListing);
  let running = false;
  let terminated = false;
  return {
    createOutputs: () => {
      return new import_runtime.Outputs(initResult.outputVarIds, initResult.startTime, initResult.endTime, initResult.saveFreq);
    },
    runModel: async (inputs, outputs, options) => {
      if (terminated) {
        throw new Error("Async model runner has already been terminated");
      } else if (running) {
        throw new Error("Async model runner only supports one `runModel` call at a time");
      } else {
        running = true;
      }
      params2.updateFromParams(inputs, outputs, options);
      let ioBuffer;
      try {
        ioBuffer = await modelWorker2.runModel((0, import_threads.Transfer)(params2.getEncodedBuffer()));
      } finally {
        running = false;
      }
      params2.updateFromEncodedBuffer(ioBuffer);
      params2.finalizeOutputs(outputs);
      return outputs;
    },
    terminate: () => {
      if (terminated) {
        return Promise.resolve();
      } else {
        terminated = true;
        return import_threads.Thread.terminate(modelWorker2);
      }
    }
  };
}

// src/worker.ts
var import_worker = require("threads/worker");
var import_runtime2 = require("@sdeverywhere/runtime");
var initGeneratedModel;
var runnableModel;
var params = new import_runtime2.BufferedRunModelParams();
var modelWorker = {
  async initModel() {
    if (runnableModel) {
      throw new Error("RunnableModel was already initialized");
    }
    const generatedModel = await initGeneratedModel();
    runnableModel = (0, import_runtime2.createRunnableModel)(generatedModel);
    return {
      outputVarIds: runnableModel.outputVarIds,
      modelListing: runnableModel.modelListing,
      startTime: runnableModel.startTime,
      endTime: runnableModel.endTime,
      saveFreq: runnableModel.saveFreq,
      outputRowLength: runnableModel.numSavePoints
    };
  },
  runModel(ioBuffer) {
    if (!runnableModel) {
      throw new Error("RunnableModel must be initialized before running the model in worker");
    }
    params.updateFromEncodedBuffer(ioBuffer);
    runnableModel.runModel(params);
    return (0, import_worker.Transfer)(ioBuffer);
  }
};
function exposeModelWorker(init) {
  initGeneratedModel = init;
  (0, import_worker.expose)(modelWorker);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  exposeModelWorker,
  spawnAsyncModelRunner
});
