// antlr4-vensim/parser/ModelVisitor.js
// Generated from grammar/Model.g4 by ANTLR 4.12.0
// jshint ignore: start
import antlr4 from 'antlr4';

// This class defines a complete generic visitor for a parse tree produced by ModelParser.

export default class ModelVisitor extends antlr4.tree.ParseTreeVisitor {

	// Visit a parse tree produced by ModelParser#model.
	visitModel(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#subscriptRange.
	visitSubscriptRange(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#subscriptDefList.
	visitSubscriptDefList(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#subscriptSequence.
	visitSubscriptSequence(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#subscriptMappingList.
	visitSubscriptMappingList(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#subscriptMapping.
	visitSubscriptMapping(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#equation.
	visitEquation(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#lhs.
	visitLhs(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Call.
	visitCall(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Or.
	visitOr(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Keyword.
	visitKeyword(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#MulDiv.
	visitMulDiv(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#AddSub.
	visitAddSub(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Var.
	visitVar(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Parens.
	visitParens(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Const.
	visitConst(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Relational.
	visitRelational(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#LookupCall.
	visitLookupCall(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Not.
	visitNot(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Negative.
	visitNegative(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Positive.
	visitPositive(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#And.
	visitAnd(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Equality.
	visitEquality(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#LookupArg.
	visitLookupArg(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#Power.
	visitPower(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#exprList.
	visitExprList(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#subscriptList.
	visitSubscriptList(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#lookup.
	visitLookup(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#lookupRange.
	visitLookupRange(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#lookupPointList.
	visitLookupPointList(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#lookupPoint.
	visitLookupPoint(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by ModelParser#constList.
	visitConstList(ctx) {
	  return this.visitChildren(ctx);
	}



}
